src/cpp/commandreply.cpp

```diff
@@ -30,236 +30,294 @@
 
 using namespace SmartRedis;
 
-CommandReply::CommandReply(redisReply* reply)
+CommandReply::CommandReply(const CommandReply& reply)
 {
-  this->_uptr_reply = 0;
-  this->_reply = reply;
+    if (this != &reply) {
+        _uptr_reply = RedisReplyUPtr(deep_clone_reply(reply._uptr_reply.get()),
+                                     sw::redis::ReplyDeleter());
+        _reply = _uptr_reply.get();
+    }
+}
+
+CommandReply::CommandReply(const redisReply* reply)
+{
+    _uptr_reply = RedisReplyUPtr(deep_clone_reply(reply),
+                                 sw::redis::ReplyDeleter());
+    _reply = _uptr_reply.get();
 }
 
 CommandReply::CommandReply(RedisReplyUPtr&& reply)
 {
-  this->_uptr_reply = std::move(reply);
-  this->_reply = this->_uptr_reply.get();
-  return;
+    _uptr_reply = std::move(reply);
+    _reply = _uptr_reply.get();
 }
 
 CommandReply::CommandReply(redisReply*&& reply)
 {
-  this->_uptr_reply = 0;
-  this->_reply = std::move(reply);
-  return;
+    _uptr_reply = NULL;
+    _reply = std::move(reply);
 }
 
 CommandReply::CommandReply(CommandReply&& reply)
 {
-  this->_uptr_reply = std::move(reply._uptr_reply);
-  this->_reply = this->_uptr_reply.get();
-  return;
+    if (this != &reply) {
+        _uptr_reply = std::move(reply._uptr_reply);
+        _reply = _uptr_reply.get();
+    }
+}
+
+CommandReply& CommandReply::operator=(const CommandReply& reply)
```

Comment:
I might be wrong on this (maybe we can write a quick test), but for some of these``deep_clone_reply()`` calls is there a potential problem with scenarios of shallowed cloned ``CommandReply`` objects later being copied?  I sketched out a scenario below:

If I have ``CommandReply`` of type ``REDIS_REPLY_ARRAY `` returned from a redis++ call:
```c++
CommandReply first_reply = client.some_redis_call();
// first_ptr._uptr_reply = u_ptr_1
// first_ptr._reply = u_ptr_1.get()
```

Then if I call the ``operator[]`` function:

```c++
CommandReply sub_reply = first_reply[0];
//sub_reply._uptr_reply = NULL
//sub_reply._reply = u_ptr_1.get()
```

Then if I call the copy constructor:
```
CommandReply copy_sub_reply(sub_reply);
```
Are ``_uptr_reply`` and ``_reply`` of ``copy_sub_reply`` both ``NULL``?